app/src/lib/markdown-filters/emoji-filter.ts

```diff
@@ -0,0 +1,138 @@
+import { fatalError } from '../fatal-error'
+import { INodeFilter } from './node-filter'
+import * as FSE from 'fs-extra'
+import { escapeRegExp } from '../helpers/regex'
+import uri2path from 'file-uri-to-path'
+
+/**
+ * The Emoji Markdown filter will take a text node and create multiple text and
+ * image nodes by inserting emoji images using base64 data uri where emoji
+ * references are in the text node.
+ *
+ * Example: A text node of "That is great! :+1: Good Job!"
+ * Becomes three nodes: "That is great! ",<img src="data uri for :+1:>, " Good Job!"
+ *
+ * Notes: We are taking the emoji file paths and creating the base 64 data URI
+ * because this is to be injected into a sandboxed markdown parser were we will
+ * no longer have access to the local file paths.
+ */
+export class EmojiFilter implements INodeFilter {
+  private readonly emojiRegex: RegExp
+  private readonly emojiFilePath: Map<string, string>
+  private readonly emojiBase64URICache: Map<string, string> = new Map()
+
+  /**
+   * @param emoji Map from the emoji ref (e.g., :+1:) to the image's local path.
+   */
+  public constructor(emojiFilePath: Map<string, string>) {
+    this.emojiFilePath = emojiFilePath
+    this.emojiRegex = this.buildEmojiRegExp(emojiFilePath)
+  }
+
+  /**
+   * Emoji filter iterates on all text nodes that are not inside a pre or code tag.
+   */
+  public createFilterTreeWalker(doc: Document): TreeWalker {
+    return doc.createTreeWalker(doc, NodeFilter.SHOW_TEXT, {
+      acceptNode: function (node) {
+        return node.parentNode !== null &&
+          ['CODE', 'PRE'].includes(node.parentNode.nodeName)
+          ? NodeFilter.FILTER_SKIP
+          : NodeFilter.FILTER_ACCEPT
+      },
+    })
+  }
+
+  /**
+   * Takes a text node and creates multiple text and image nodes by inserting
+   * emoji image nodes using base64 data uri where emoji references are.
+   *
+   * Example: A text node of "That is great! :+1: Good Job!" Becomes three
+   * nodes: ["That is great! ",<img src="data uri for :+1:>, " Good Job!"]
+   */
+  public async filter(node: Node): Promise<ReadonlyArray<Node> | null> {
+    if (!(node instanceof Text)) {
+      fatalError(
+        'Emoji filter requires text nodes; otherwise we may inadvertently replace non text elements.'
+      )
+    }
+
+    if (node.textContent === null || !node.textContent.includes(':')) {
+      return null
+    }
+
+    let text = node.textContent
```

Comment:
Do we really need to throw for this scenario or could we simply return null here?

Perhaps we could combine a few of these assertions into one:

```suggestion
    let text = node.textContent
    if (node.nodeType !== node.TEXT_NODE || text === null) {
      return null
    }
```

Reply:
I was thinking along the lines of if a dev in the future used this `filter` method without being in conjunction with it's tree walker creation method, then, the fatalError would explain to them quickly as to why giving this a non text node could be problematic. But, I moved that warning up into the method doc which should be sufficient for a scenario that "should be" unlikely. :D